clients/src/main/java/org/apache/kafka/common/security/auth/SaslExtensions.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.security.auth;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A simple immutable value object class holding customizable SASL extensions
+ */
+public class SaslExtensions {
+    protected Map<String, String> extensionsMap;
+
+    public SaslExtensions(Map<String, String> extensionsMap) {
+        this.extensionsMap = Collections.unmodifiableMap(new HashMap<>(extensionsMap));
+    }
+
+    /**
+     * Returns an <strong>immutable</strong> map of the extension names and their values
+     */
+    public Map<String, String> map() {
```

Comment:
Personally, I would get rid of this and use `extensionValue` and `extensionNames`. Otherwise, as @rondagostino said below, we should remove `extensionValue`.

Reply:
`OAuthBearerClientInitialResponse` uses this to iterate over all values and build the extensions string using `Utils.mkString`. If we removed `map()`, we would need to iterate through `extensionNames`, fetch and validate each value one by one.
We would also need to rebuild the map in `#extensionsMessage()` so that we could call `Utils.mkString`.
This is all more complex to write and slower to execute, thus I believe we should keep `map()`.

I do not understand what is inherently wrong with having an `extensionValue` method. It keeps it consistent with `ScramExtensions`' usage, does not bloat the API (it's a single method) and offers a more concise and readable way to fetch a value from the extensions. 

Can you elaborate why you believe we should remove `extensionValue()` ?

Reply:
Come to think of it, I'll outright remove `extensionNames` and use `map().keySet()`. Did not realize `ScramExtensions` is not a public class